File: tests/compile-fail/ForeignNullableGenericImageSiblingCrossing.cs
using System;
using System.Collections.Generic;

// TWO REAL SLOTS, ONE PHYSICAL KOTLIN SIGNATURE. The erased image of `Take(List<int?>)` is `Take(List<object>)`,
// and here a sibling declares that outright — so both Kotlin overrides, `Take(xs: List<Int?>)` and
// `Take(ys: List<Any?>)`, emit the same CLR member. Only the second one legitimately fills a slot; the first
// belongs to the `List<int?>` one, which no Kotlin body can fill, and the CLR would bind it to the `object` slot
// while `Take(List<int?>)` kept the base implementation — a silently wrong answer rather than a diagnostic.
//
// The accepted half is driven in tests/interop (KotlinImageSiblingOverride), which is what keeps this refusal from
// swinging back into rejecting the sibling override.
namespace plainnet
{
    public class ImageSiblingBase
    {
        public virtual string Take(List<int?> xs) => "net-q";

        public virtual string Take(List<object> ys) => "net-o";
    }
}

File: tests/compile-fail/ForeignNullableGenericImageSiblingCrossing.kt
// #86 — the crossing override whose ERASED IMAGE is a sibling slot's real signature.
//
// `Take(xs: List<Int?>)` is the `List<int?>` slot's override and physically states `Take(List<object>)`, which the
// base also declares for real. Deciding ownership physically let the sibling discharge the crossing: this compiled,
// and the CLR bound the emitted body to the `object` slot while a call through `Take(List<int?>)` ran the base's
// own implementation — the silent wrong answer this refusal exists to prevent.
//
// Which source slot a body belongs to is answered by the fact the erasure recorded on the declaration: this
// parameter carries its pre-erasure `List<Int?>` on `[KotlinNullableGeneric]`, and the sibling override of
// `List<Any?>` — accepted, and driven in tests/interop — carries nothing there.
import plainnet.ImageSiblingBase
import System.Collections.Generic.List

class CImageSibling : ImageSiblingBase() {
    override fun Take(xs: List<Int?>): String = "kt-q"
}

fun main() {
    println(CImageSibling().toString().substring(0, 2))
}
